[EVM] Contract deploy change return value

sideninja opened the pull request:
Closes: #5505 

**❗️ API Breaking Change**

Change in the EVM COA deploy function return value, previously it returned a byte array containing the newly deployed contract address, now it returns the result type that is consistent with `run` and `call` functions. The address is stored as result data.

ramtinms commented on fvm/evm/handler/handler.go:
```diff
@@ -558,10 +558,10 @@ func (a *Account) transfer(to types.Address, balance types.Balance) (*types.Resu
-func (a *Account) Deploy(code types.Code, gaslimit types.GasLimit, balance types.Balance) types.Address {
+func (a *Account) Deploy(code types.Code, gaslimit types.GasLimit, balance types.Balance) *types.ResultSummary {
 	res, err := a.deploy(code, gaslimit, balance)
 	panicOnErrorOrInvalidOrFailedState(res, err)
```
This has to change to Error only.

ramtinms commented on fvm/evm/stdlib/contract.go:
```diff
@@ -1639,9 +1663,11 @@ func newInternalEVMTypeDeployFunction(
 
-			return EVMAddressToAddressBytesArrayValue(inter, address)
+			// assign deployed contract address to return value since it's the only result of the deploy
+			result.ReturnedValue = result.DeployedContractAddress[:]
```
I feel this logic should be part of the emulator.

m-Peter replied:
IMHO, this might complicate things, because EVM events rely on `types.Result` to generate their payload, e.g.

```go
cadence.NewUInt8(p.Result.TxType),
cadence.NewUInt64(p.Result.GasConsumed),
cadence.String(p.Result.DeployedContractAddress.String()),
cadence.String(hex.EncodeToString(p.Result.ReturnedValue)),
```

Special casing this assignment here, solely for the purpose of populating `EVM.Result` struct seems like a good way to go.

ramtinms commented:
I think we also need to update the tests in the handler and ideally an integration test in the evm_test

m-Peter commented on fvm/evm/handler/handler.go:
```diff
@@ -558,10 +558,10 @@ func (a *Account) transfer(to types.Address, balance types.Balance) (*types.Resu
 // Deploy deploys a contract to the EVM environment
 // the new deployed contract would be at the returned address and
```
I believe this line also needs an update, to describe that we now return a `types.ResultSummary` instead of a `types.Address`.

turbolent commented:
Nice! Please also amend the FLIP accordingly

sideninja commented:
> Nice! Please also amend the FLIP accordingly

Good point, updated in e0122e0378d88214648e6ffb3baa8995bd397610

m-Peter commented on fvm/evm/emulator/emulator.go:
```diff
@@ -384,6 +384,7 @@ func (proc *procedure) deployAt(
 
 	proc.state.SetCode(addr, ret)
 	res.DeployedContractAddress = to
+	res.ReturnedValue = to[:] // assign address to return value since it's the only returned value
```
I have one concern about making this change here. Right now, a contract deployment results in the following `evm.TransactionExecuted` event payload:

```cadence
evm.TransactionExecuted(
    blockHeight: 3,
    blockHash: "0xc9a60a5503928c637fbed960b8224e8786f147d1e9a4b6245527f50df0e7fbad",
    transactionHash: "0xe414f90fea2aebd75e8c0b3b6a4a0c9928e86c16ea724343d884f40bfe2c4c6b",
    transaction: "f9015880808301e8488080b901086060604052341561000f57600080fd5b60eb8061001d6000396000f300606060405260043610603f576000357c0100000000000000000000000000000000000000000000000000000000900463ffffffff168063c6888fa1146044575b600080fd5b3415604e57600080fd5b606260048080359060200190919050506078565b6040518082815260200191505060405180910390f35b60007f24abdb5865df5079dcc5ac590ff6f01d5c16edbc5fab4e195d9febd1114503da600783026040518082815260200191505060405180910390a16007820290509190505600a165627a7a7230582040383f19d9f65246752244189b02f56e8d0980ed44e7a56c0b200458caad20bb002982052fa09c05a7389284dc02b356ec7dee8a023c5efd3a9d844fa3c481882684b0640866a057e96d0a71a857ed509bb2b7333e78b2408574b8cc7f51238f25c58812662653",
    failed: false,
    vmError: "",
    transactionType: 0,
    gasConsumed: 103856,
    deployedContractAddress: "0x99466ED2E37B892A2Ee3E9CD55a98b68f5735db2",
    returnedValue: "606060405260043610603f576000357c0100000000000000000000000000000000000000000000000000000000900463ffffffff168063c6888fa1146044575b600080fd5b3415604e57600080fd5b606260048080359060200190919050506078565b6040518082815260200191505060405180910390f35b60007f24abdb5865df5079dcc5ac590ff6f01d5c16edbc5fab4e195d9febd1114503da600783026040518082815260200191505060405180910390a16007820290509190505600a165627a7a7230582040383f19d9f65246752244189b02f56e8d0980ed44e7a56c0b200458caad20bb0029",
    logs: ""
)
```

Where `returnedValue` holds the executed byte-code, basically what should be returned by `eth_getCode` for the deployed contract address. 

With `res.ReturnedValue = to[:]`, we will basically returned the deployed contract address for both `deployedContractAddress` & `returnedValue`. 

This results in less data being available in the emitted event. For example, the Gateway can use the current event payload and index the executed byte-code for each address.

sideninja replied:
That is true and I would say the only way to resolve this would to add the deployed address value to the `EVM.Result` type, the other way is to revert to overwrite the `ReturnedValue` in the contract handler no in emulator. cc @ramtinms

sideninja replied:
@m-Peter I added deployed contract address as the new field on the result type

turbolent replied:
Left a question in the FLIP: https://github.com/onflow/flips/pull/225#discussion_r1561285117

sideninja commented:
> @sideninja Thanks for the update of the FLIP! Maybe an accident, but [onflow/flips@e0122e0](https://github.com/onflow/flips/commit/e0122e0378d88214648e6ffb3baa8995bd397610) also reverted all the Cadence 1.0 changes (access modifiers with entitlements, view annotations, etc.). Could you please revert those?

Oh right, the master flow-go is not Cadence 1.0. Fixed, thanks for flagging! https://github.com/onflow/flips/pull/225/commits/3f8c07170dc405d914100dc0073c2ac161f7cd39

sideninja commented:
@ramtinms yes it is, and the field has been added: https://github.com/onflow/flow-go/pull/5606/files#diff-02837edc597afad5bd8f7d1962c129bded25adcc21a9889382c7aaad9a0e5b34R180

sisyphusSmiling commented on fvm/evm/stdlib/contract.cdc:
```diff
@@ -168,21 +168,29 @@ contract EVM {
             gasUsed: UInt64,
-            data: [UInt8]
+            data: [UInt8],
+            contractAddress: [UInt8; 20]
```
I'm a bit late to the party here - what is this value for `CadenceOwnedAccount.call` - I'd imagine a `0` array since the field isn't optional?

sideninja replied:
I will make it optional, it's a good suggestion

sideninja commented on fvm/evm/stdlib/contract.cdc:
```diff
@@ -168,21 +168,29 @@ contract EVM {
+        /// if the transaction caused such a deployment
+        /// otherwise the value is empty.
+        access(all)
+        let deployedContractAddress: [UInt8; 20]
```
very good point!

sideninja commented:
The deployed contract field on `Result` type was changed to be an optional as per @sisyphusSmiling comment. 
cc @ramtinms

sisyphusSmiling reviewed (commented):
`Result` looks good, thanks for the updates!
